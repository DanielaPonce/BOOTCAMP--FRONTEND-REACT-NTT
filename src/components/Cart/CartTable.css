.cart-table-container {
	width: 100%;
	overflow-x: scroll;
}

.cart-table {
	width: 100%;
	margin-top: 1rem;
	overflow-x: scroll;
}

.cart-table th,
.cart-table td {
	padding: 4px 1rem;
	border: unset;
	background-color: white;

}


.cart-table th {
	background-color: var(--secondary-color);
	color: var(--white);
}

.cart-table td button {
	padding: 2px 4px;
	border: 1px solid var(--border-color);
	border-radius: 4px;
	cursor: pointer;
}

td img {
	width: 5rem;
	height: 5rem;
	object-fit: contain;
	margin: 0 auto;
	display: flex;
}

td .quantity-content, .remove-button-content {
	display: flex;
	align-items: center;
	justify-content: center;
	gap: 8px;
}

.remove-button-content button {
	background-color: var(--button-error);
	color: var(--white);
	cursor: pointer;
}