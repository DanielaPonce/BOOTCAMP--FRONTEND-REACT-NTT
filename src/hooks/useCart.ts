import { useEffect } from 'react';
import { useCartContext } from '../context/CartContext';
import { Product } from '../models/product.types';
import { Actions } from '../utils/actions';
import { LocalStorageConstants } from '../utils/localStorageConstants';

export const useCart = () => {
	const { state, dispatch } = useCartContext();

	useEffect(() => {
		localStorage.setItem(
			LocalStorageConstants.CartKey,
			JSON.stringify(state)
		);
	}, [state]);

	const addToCart = (product: Product) =>
		dispatch({ type: Actions.ADD_TO_CART_TYPE, payload: product });

	const removeFromCart = (id: number) =>
		dispatch({ type: Actions.REMOVE_FROM_CART_TYPE, payload: id });

	const updateQuantity = (id: number, quantity: number) =>
		dispatch({
			type: Actions.UPDATE_QUANTITY_TYPE,
			payload: { id, quantity }
		});

	const clearCart = () => dispatch({ type: Actions.CLEAR_CART_TYPE });

	return {
		cart: state.cart,
		total: state.total,
		itemCount: state.itemCount,
		addToCart,
		removeFromCart,
		updateQuantity,
		clearCart
	};
};
