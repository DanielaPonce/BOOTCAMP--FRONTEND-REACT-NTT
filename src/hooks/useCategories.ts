import { useEffect, useState } from 'react';
import { fetchAllCategories } from '../api/category.api';
import { Category } from '../models/category.types';

export const useCategories = () => {
	const [categories, setCategories] = useState<Category[]>([]);

	useEffect(() => {
		fetchAllCategories().then((data) => setCategories(data));
	}, []);

	return { categories };
};
