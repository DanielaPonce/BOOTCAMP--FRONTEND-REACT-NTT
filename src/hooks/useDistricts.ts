import { useEffect, useState } from 'react';

interface District {
	id: number;
	name: string;
}

export const useDistricts = () => {
	const [districts, setDistricts] = useState<District[]>([]);

	useEffect(() => {
		const fetchDistricts = async () => {
			const response = await fetch('/src/assets/mock/districts.json');
			const data: District[] = await response.json();
			setDistricts(data);
		};

		fetchDistricts();
	}, []);

	return districts;
};
