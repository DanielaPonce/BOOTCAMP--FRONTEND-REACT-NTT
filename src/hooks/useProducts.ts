import { useEffect, useState } from 'react';
import { fetchAllProducts, fetchProductsByCategory } from '../api/product.api';
import { Product } from '../models/product.types';
import { Constants } from '../utils/constants';

export const useProducts = (category: string, search: string) => {
	const [allProducts, setAllProducts] = useState<Product[]>([]);
	const [filteredProducts, setFilteredProducts] = useState<Product[]>([]);
	const [isLoading, setIsLoading] = useState(true);
	const [currentPage, setCurrentPage] = useState(1);

	const itemsPerPage = 8;

	const loadProducts = async () => {
		setIsLoading(true);
		try {
			let products: Product[] = [];
			if (category === Constants.defaultSelectValue) {
				products = await fetchAllProducts();
			} else {
				products = await fetchProductsByCategory(category);
			}
			setAllProducts(products);
			setFilteredProducts(products);
			setCurrentPage(1);
		} catch (error) {
			console.log('Error al cargar los productos:', error);
		} finally {
			setIsLoading(false);
		}
	};

	useEffect(() => {
		loadProducts();
	}, [category]);

	useEffect(() => {
		const filtered = allProducts.filter((product) =>
			product.title.toLowerCase().includes(search.toLowerCase())
		);
		setFilteredProducts(filtered);
		setCurrentPage(1);
	}, [search, allProducts]);

	const paginatedProducts = filteredProducts.slice(
		(currentPage - 1) * itemsPerPage,
		currentPage * itemsPerPage
	);

	const totalPages = Math.ceil(filteredProducts.length / itemsPerPage);

	return {
		products: paginatedProducts,
		isLoading,
		totalPages,
		currentPage,
		setCurrentPage,
		showPagination: totalPages !== 1
	};
};
