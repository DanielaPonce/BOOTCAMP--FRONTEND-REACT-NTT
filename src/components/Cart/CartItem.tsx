import { FC } from 'react';
import { Product } from '../../models/product.types';
import { CartItemType } from '../../context/CartContext';
import { useCart } from '../../hooks/useCart';

interface CartItemProps {
	item: CartItemType;
}

const CartItem: FC<CartItemProps> = ({ item }) => {
	const { removeFromCart, updateQuantity } = useCart();

	const handleIncrease = () => {
		updateQuantity(item.id, item.quantity + 1);
	};

	const handleDecrease = () => {
		if (item.quantity > 1) {
			updateQuantity(item.id, item.quantity - 1);
		} else {
			removeFromCart(item.id);
		}
	};

	return (
		<tr>
			<td>
				<img
					src={item.thumbnail}
					alt={item.title}
					className="product-image"
				/>
			</td>
			<td>{item.title}</td>
			<td>
				<button onClick={handleDecrease}>-</button>
				<span>{item.quantity}</span>
				<button onClick={handleIncrease}>+</button>
			</td>
			<td>
				<button
					onClick={() => removeFromCart(item.id)}
					className="remove-button"
				>
					Eliminar
				</button>
			</td>
		</tr>
	);
};

export default CartItem;
