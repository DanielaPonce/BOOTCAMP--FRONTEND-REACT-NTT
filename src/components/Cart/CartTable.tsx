import { useCart } from '../../hooks/useCart';
import { FC } from 'react';
import CartItem from './CartItem';

const CartTable: FC = () => {
	const { cart } = useCart();

	return (
		<table className="cart-table">
			<thead>
				<tr>
					<th>Producto</th>
					<th>Nombre</th>
					<th>Cantidad</th>
					<th>Eliminar</th>
				</tr>
			</thead>
			<tbody>
				{cart.map((item) => (
					<CartItem key={item.id} item={item} />
				))}
			</tbody>
		</table>
	);
};

export default CartTable;
