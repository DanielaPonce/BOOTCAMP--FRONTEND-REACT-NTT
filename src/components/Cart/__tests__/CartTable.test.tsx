import { jest } from '@jest/globals';
import { render, screen } from '@testing-library/react';
import { useCart } from '../../../hooks/useCart';
import CartItem from '../CartItem';
import CartTable from '../CartTable';

jest.mock('../../../hooks/useCart');
jest.mock('../CartItem', () => jest.fn(() => <tr data-testid="cart-item" />));

describe('CartTable Component', () => {
	const mockCart = [
		{ id: 1, name: 'Producto 1', price: 10.5, quantity: 2 },
		{ id: 2, name: 'Producto 2', price: 20.0, quantity: 1 }
	];

	beforeEach(() => {
		(useCart as jest.Mock).mockReturnValue({ cart: mockCart });
		(CartItem as jest.Mock).mockClear();
	});

	test('renders correctly', () => {
		render(<CartTable />);

		const table = screen.getByRole('table');
		expect(table).toBeInTheDocument();

		const columnHeaders = [
			'Producto',
			'Nombre',
			'Precio',
			'Cantidad',
			'Eliminar'
		];
		columnHeaders.forEach((header) => {
			expect(screen.getByText(header)).toBeInTheDocument();
		});
	});

	test('renders the correct number of CartItem components', () => {
		render(<CartTable />);

		const cartItems = screen.getAllByTestId('cart-item');
		expect(cartItems.length).toBe(mockCart.length);
		expect(CartItem).toHaveBeenCalledTimes(mockCart.length);

		mockCart.forEach((item, index) => {
			expect(CartItem).toHaveBeenNthCalledWith(index + 1, { item }, {});
		});
	});

	test('renders an empty table when the cart is empty', () => {
		(useCart as jest.Mock).mockReturnValue({ cart: [] });

		render(<CartTable />);

		const table = screen.getByRole('table');
		expect(table).toBeInTheDocument();

		const cartItems = screen.queryByTestId('cart-item');
		expect(cartItems).toBeNull();
	});

	test('matches snapshot', () => {
		const { asFragment } = render(<CartTable />);
		expect(asFragment()).toMatchSnapshot();
	});
});
