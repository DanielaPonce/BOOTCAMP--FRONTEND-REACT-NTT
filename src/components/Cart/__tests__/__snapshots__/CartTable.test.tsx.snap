// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartTable Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="cart-table-container"
  >
    <table
      class="cart-table"
    >
      <thead>
        <tr>
          <th>
            Producto
          </th>
          <th>
            Nombre
          </th>
          <th>
            Precio
          </th>
          <th>
            Cantidad
          </th>
          <th>
            Eliminar
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          data-testid="cart-item"
        />
        <tr
          data-testid="cart-item"
        />
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
